[TLVB-111] 사용자 정보 수정 API 구현

nasaoreo opened the pull request:
## 👩‍💻 요구 사항과 구현 내용
사용자 정보 수정 API를 구현했습니다.

사용자 정보는 세가지 경우로 수정할 수 있습니다.

1) 닉네임만 수정
2) 비밀번호만 수정
3) 닉네임과 비밀번호를 같이 수정

## ✅ 피드백 반영사항

## 🤔 PR 포인트 & 궁금한 점

## 관련 이슈 
TLVB-111

nasaoreo committed:
chore: 테스트 코드에서 slf4j 사용을 위한 dependency 추가

nasaoreo committed:
feat: 사용자 정보 수정 기능 추가

nasaoreo commented on src/main/java/kdt/prgrms/kazedon/everevent/domain/user/User.java:
```diff
@@ -77,6 +78,15 @@ public User(SignUpRequest request){
     addAuthority(new Authority(this, "ROLE_USER"));
   }
 
+  public void changePassword(String password, PasswordEncoder passwordEncoder){
```
회원가입 코드를 구현할 때, UserService에서 RequestDto의 패스워드 필드에 대해 인코딩을 하셨더라구요! User 생성자에서는 인코딩된 password 값만 입력 받구요!

[UserService.java]
``` java
public Long signUp(SignUpRequest request) {
    return userRepository.save(new User(encodingPassword(request))).getId();
  }

public SignUpRequest encodingPassword(SignUpRequest request) {
    request.encodingPassword(passwordEncoder.encode(request.getPassword()));
    return request;
  }
```

저는 password를 변경할 때, passwordEncoder를 주입받아 사용하는 방법으로 구현해봤는데, 의견 부탁드려요!
인코딩되기 전의 password에 대해 유효성 검사를 하고, 인코딩을 할 수 있다는 장점이 있어요~!

dia312 replied:
인코딩되기 전의 유효성 검사를 서비스 레이어에서 하면 어떨까요?

nasaoreo commented on src/main/java/kdt/prgrms/kazedon/everevent/exception/ErrorMessage.java:
```diff
@@ -7,7 +7,7 @@
   DUPLICATE_NOT_FAVORITE_MARKET("아이디 {0}은 이미 삭제한 즐겨찾기 항목입니다."),
   INVALID_EMAIL_FORMAT("이메일 {0}은 잘못된 형식입니다."),
-  INVALID_PASSWORD_FORMAT("비밀번호 {0}은 잘못된 형식입니다."),
+  INVALID_PASSWORD_FORMAT("비밀번호는 1자 이상 100자 이하로 입력되어야 합니다."),
```
이 Message는 응답 바디에도 들어가는 부분이기 때문에, 비밀번호를 어떻게 입력했는지 명시하지 않고 어떻게 잘못된 형식인지를 기술하도록 변경했습니다!

nasaoreo committed:
refactor: UserUpdateRequest 멤버변수의 한정자 추가

ksy90101 commented on build.gradle:
```diff
@@ -40,7 +40,8 @@ dependencies {
     testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc'
     asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor:2.0.2.RELEASE'
-    implementation 'org.slf4j:slf4j-api:1.7.32'
+    testCompileOnly 'org.projectlombok:lombok'
+    testAnnotationProcessor 'org.projectlombok:lombok'
```
이걸 추가한 이유가 뭔가요?

nasaoreo replied:
> 이걸 추가한 이유가 뭔가요?

테스트 코드에서 로깅을 하기 위해 추가했습니다.

ksy90101 replied:
테스트 코드에서 로깅하는 이유가 무엇인가요?

nasaoreo replied:
> 테스트 코드에서 로깅하는 이유가 무엇인가요?

사용자 정보 수정 기능을 테스트하면서 로깅을 추가했구요.
수정을 했을 때 assert 문으로 확인도 하지만 가시적으로 어떻게 바뀌는지 한번 더 확인하기 위해 추가했어요.
개인적인 견해이지만, 로그를 찍어보면서 패스워드는 잘 인코딩 되었는지, 변경되기 전/후의 정보를 직접 볼 수 있어서 테스트 결과에 더 믿음이 가더라구요.

nasaoreo replied:
> 테스트 코드에서 로깅하는 이유가 무엇인가요?

제가 이 부분에 대해서 좀 더 고민을 해봤는데요, 테스트 코드에서 로그를 찍는 것 보다는 서비스 코드에서 로그를 찍는다면 굳이 테스트 코드에서 따로 로그를 찍을 필요는 없겠네요! 그런데 별도로 테스트 코드에서 원하는 결과를 확인하고 싶을 때에는 어떻게 처리해야할 지 궁금하네요!  좋은 질문 주셔서 감사해요~!

ksy90101 replied:
테스트는 기본적으로 자동화가 원칙입니다.
FIRST라는 원칙이 있는데요. 여기서 S가
`Self-validating, 스스로 검증 가능해야 한다`
이런 뜻입니다.

제 생각은 Log를 찍는다는 것 자체가 개발자가 테스트를 확인해야 하고, 그건 테스트가 스스로 검증을 하는게 아니지 않을까요?
테스트를 실행시켰을때 통과가 되면 그것으로 테스트는 정상적으로 되었다고 믿고 가야합니다. (믿고 가다보면 잘못된 테스트가 될수도 있기 때문에 테스트 케이스가 사실 중요하긴 합니다 :) )

ksy90101 commented on src/main/java/kdt/prgrms/kazedon/everevent/domain/user/repository/UserRepository.java:
```diff
@@ -12,8 +13,8 @@
 
   @Query("select (count(u) > 0) from User u where u.email = :email")
-  boolean existsByEmail(String email);
+  boolean existsByEmail(@Param("email") String email);
 
   @Query("select (count(u) > 0) from User u where u.nickname = :nickname")
-  boolean existsByNickname(String nickname);
+  boolean existsByNickname(@Param("nickname") String nickname);
```
1. 쿼리 작성을 따로 해야 하나요?
https://www.baeldung.com/spring-data-exists-query#using-a-derived-query-method

nasaoreo replied:
이 부분은 제가 잘 살펴보지 않았네요! 불필요한 사용자 정의 쿼리는 삭제했습니다!

nasaoreo committed:
refactor: 불필요한 사용자 정의 쿼리문 삭제

dia312 reviewed (approved):
고생하셨습니다 진아님~~~!!! 👍

Heejeong01110 commented on src/main/java/kdt/prgrms/kazedon/everevent/domain/user/dto/UserUpdateRequest.java:
```diff
@@ -0,0 +1,24 @@
+    @Size(max = 20)
+    private String nickname;
+
+    @Size(max = 100)
+    private String password;
```
이부분 프론트분들이랑 얘기한대로 6자 이상 영어 소문자와 숫자로 @pattern 사용해서 유효성 검사 하면 좋을 것 같아요~!!
추가로 회원가입하는 부분도 같이 수정 해주시면 감사하겠습니다 ㅎ.ㅎ

nasaoreo committed:
refactor: 비밀번호 검증 패턴 추가, 패스워드 인코딩 코드 리팩토링

- 패스워드 검증은 DTO 단에서 수행 (인코딩 되기 전의 패스워드 검증)
- 패스워드 인코딩은 UserService에서 수행
- User 클래스에서 불필요한 검증 코드 삭제
- SignUpRequest 를 이용해서 User 객체 생성 시 인코딩된 패스워드도 입력받도록 수정

Heejeong01110 commented on src/main/java/kdt/prgrms/kazedon/everevent/domain/user/dto/UserUpdateRequest.java:
```diff
@@ -14,11 +16,13 @@
     private String nickname;
 
     @Size(max = 100)
+    @Pattern(regexp = "^(?=.*[0-9])(?=.*[a-zA-Z]).{6,100}$")
```
SignUpRequest에 있는 정규식이랑 다른거 같습니다!

nasaoreo replied:
> SignUpRequest에 있는 정규식이랑 다른거 같습니다!

수정했어요~!

Heejeong01110 commented on src/main/java/kdt/prgrms/kazedon/everevent/service/UserService.java:
```diff
@@ -10,10 +10,12 @@
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
 
+@Slf4j
```
로그 더이상 사용하지 않으면 지워도 될 것 같네용~!

nasaoreo committed:
fix: 비밀번호 검증 정규식 변경
